cmd/adapter/adapter.go

```diff
@@ -0,0 +1,430 @@
+		}
+		ad.logger.Warn("Serving API stopped.")
+		ad.waitWg.Done()
+	}()
+}
+
+func (ad *Adapter) serveAPI(ctx context.Context) *http.Server {
+	router := httprouter.New()
+	router.GET("/v1/log", ad.makeHandlerWithCtx(ctx, ad.logHandler))
+	router.GET("/v1/download", ad.makeHandlerWithCtx(ctx, ad.downloadHandler))
+
+	srv := &http.Server{Addr: ad.cfg.Gravity.ApiListenAddr, Handler: router}
+	go func() {
+		if err := srv.ListenAndServe(); err != http.ErrServerClosed {
+			ad.logger.Fatal("serveAPI(): %s", err)
+		}
+	}()
+	return srv
+}
+
+func (ad *Adapter) logHandler(ctx context.Context, rw http.ResponseWriter, rq *http.Request, p httprouter.Params) error {
+	ad.logger.Info("log(): Request=", rq)
+	qQuery := strings.TrimSpace(rq.URL.Query().Get("query"))
+	qLimit := strings.TrimSpace(rq.URL.Query().Get("limit"))
+
+	var (
+		err error
+	)
+
+	limit := lgLimitLines
+	if qLimit != "" {
+		limit, err = strconv.Atoi(qLimit)
+		if err != nil || limit < 0 {
+			limit, err = lgLimitLines, nil
+		}
+	}
+
+	qr := &api.QueryRequest{
+		Query: buildLql(qQuery, ad.cfg.Logrange.Partition, limit, -limit),
+		Pos:   "tail", Offset: -limit, Limit: limit,
+	}
+
+	ad.logger.Info("log(): Query=", qr.Query)
+	rs := &api.QueryResult{}
+	err = ad.lCli.Query(ctx, qr, rs)
+	if err == nil {
+		grLogEntries, err := ad.toGrLogEntries(rs.Events)
+		if err == nil {
+			bb, err := json.Marshal(grLogEntries)
+			if err == nil {
+				_, err = rw.Write(bb)
+			}
+		}
+	}
+
+	if err != nil {
+		ad.logger.Error("log(): Err=", err)
+		rw.WriteHeader(http.StatusInternalServerError)
+		_, _ = rw.Write([]byte("oops, something went wrong!\n"))
+	}
+
+	return err
+}
+
+func (ad *Adapter) toGrLogEntries(evs []*api.LogEvent) ([]string, error) {
+	gle := make([]string, 0, len(evs))
+	le := &grLogEntry{Type: "data"}
+	for _, e := range evs {
+		le.Payload = e.Message
+		bb, err := json.Marshal(le)
+		if err != nil {
+			return nil, err
+		}
+		gle = append(gle, string(bb))
+	}
+	return gle, nil
+}
+
+func (ad *Adapter) downloadHandler(ctx context.Context,
+	rw http.ResponseWriter, rq *http.Request, p httprouter.Params) error {
+
+	gzWriter := gzip.NewWriter(rw)
+	defer gzWriter.Close()
+	tarWriter := tar.NewWriter(gzWriter)
+	defer tarWriter.Close()
+
+	lql := buildLql("", ad.cfg.Logrange.Partition, dlLimitLines, -dlLimitLines)
+	ad.logger.Info("download(): Query=", lql)
+
+	var (
+		fnCntr = 0
+		err    error
+	)
+
+	buf := bytes.Buffer{}
+	err = api.Select(ctx, ad.lCli, &api.QueryRequest{Query: lql,
+		Limit: dlLimitLines}, false, func(res *api.QueryResult) {
+		if err == nil {
+			marshal(res.Events, &buf)
+			if buf.Len() > dlLimitPerFileBt {
+				err = writeTar(fnCntr, buf.Bytes(), tarWriter)
+				buf.Reset()
+				fnCntr++
+			}
+		}
+	})
+
+	if err == nil {
+		if buf.Len() > 0 {
+			err = writeTar(fnCntr, buf.Bytes(), tarWriter)
+		}
+	}
+
+	if err != nil {
+		ad.logger.Error("download(): Err=", err)
+		_ = writeTar(fnCntr, []byte("oops, something went wrong!\n"), tarWriter)
```

Comment:
I don't think reporting errors as compressed tarballs is good practice - I would stream the logs to the local file system, handle the intermediate errors and then attach the file as a tarball.